FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+- Xarray coordinate getters (see [Section 2.2.4](#224-implicit-coodinates))
+- A method that may return a new index and that will be called when one of the corresponding coordinates is dropped from the Dataset/DataArray (multi-coordinate indexes)
+- `encode()`/`decode()` methods that would allow storage-agnostic serialization and fast-path reconstruction of the underlying index object(s) (see [Section 2.8](#28-index-encoding))
+- One or more "non-standard" methods or properties that could be leveraged in Xarray 3rd-party extensions like Dataset/DataArray accessors (see [Section 2.7](#27-using-indexes-for-other-purposes))
+
+The `XarrayIndex` API has still to be defined in detail.
+
+Xarray should provide a minimal set of built-in index wrappers (this could be reduced to the indexes currently supported in Xarray, i.e., `pandas.Index` and `pandas.MultiIndex`). Other index wrappers may be implemented in 3rd-party libraries (recommended). The `XarrayIndex` base class should be part of Xarray's public API.
+
+#### 2.1.1 Index discoverability
+
+For better discoverability of Xarray-compatible indexes, Xarray could provide some mechanism to register new index wrappers, e.g., something lire [xoak's `IndexRegistry`](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.IndexRegistry.html#xoak.IndexRegistry).
+
+Additionally (or alternatively), new index wrappers may be registered via entry points like it is already the case for storage backends and maybe other backends (plotting) in the future.
+
+`XarrayIndex` subclasses may still be used directly when setting new indexes from DataArray/Dataset coordinates.
+
+### 2.2 Explicit vs. implicit index creation
+
+#### 2.2.1 Dataset/DataArray's `indexes` constructor argument
+
+The new `indexes` argument of Dataset/DataArray constructors may be used to specify which kind of index to bind to which coordinate(s). It would consist of a mapping where, for each item, the key is one coordinate name (or a sequence of coordinate names) that must be given in `coords` and the value is the type of the index to build from this (these) coordinate(s):
+
+```python
+>>> da = xr.DataArray(
+...     data=[[275.2, 273.5], [270.8, 278.6]],
+...     dims=('x', 'y'),
+...     coords={
+...         'lat': (('x', 'y'), [[45.6, 46.5], [50.2, 51.6]]),
+...         'lon': (('x', 'y'), [[5.7, 10.5], [6.2, 12.8]]),
+...     },
+...     indexes={('lat', 'lon'): SpatialIndex},
+... )
+<xarray.DataArray (x: 2, y: 2)>
+array([[275.2, 273.5],
+       [270.8, 278.6]])
+Coordinates:
+  * lat      (x, y) float64 45.6 46.5 50.2 51.6
+  * lon      (x, y) float64 5.7 10.5 6.2 12.8
+```
+
+More formally, `indexes` would accept `Mapping[CoordinateNames, IndexSpec]` where:
+
+- `CoordinateNames = Union[CoordinateName, Tuple[CoordinateName, ...]]` and `CoordinateName = Hashable`
+- `IndexSpec = Union[Type[XarrayIndex], Tuple[XarrayIndex, Dict[str, Any]], XarrayIndex]`, so that index instances or index classes + build options could be also passed
```

Comment:
I'm not sure I exactly understand what each of these options would correspond to. I guess `XarrayIndex` is the only essential one, and the others are for user convenience?

Reply:
There's a mistake: `Tuple[XarrayIndex, Dict[str, Any]]` ->  `Tuple[Type[XarrayIndex], Dict[str, Any]]`

Some examples:

- `Type[XarrayIndex]`

```python
da = xr.DataArray(
    data=...,
    dims=('x', 'y'),
    coords={'lat': (('x', 'y'), ...), 'lon': (('x', 'y'), ...)},
    indexes={('lat', 'lon'): SpatialIndex},
)
```

- `Tuple[Type[XarrayIndex], Dict[str, Any]]`

```python
>>> da = xr.DataArray(
    data=...,
    dims=('x', 'y'),
    coords={'lat': (('x', 'y'), ...), 'lon': (('x', 'y'), ...)},
    indexes={('lat', 'lon'): (SpatialIndex, {'kdtree_leafsize': 20})},
)
```

- `XarrayIndex`

```python
midx = pd.MultiIndex.from_arrays(..., names=['lvl1' , 'lvl2'])

coords_dict, index_dict = create_coords_from_index(midx, dims='x')

# index_dict == {('lvl1', lvl2'): midx}

da =  xr.DataArray(..., dims='x', coords=coords_dict, indexes=index_dict)
```